### PR TITLE
First try at moving isimplicit to declaration from method.

### PR DESCRIPTION
**Pull request by genuinelucifer**
I have yet to add a test.
.
Please see if I am going in right direction or not.
Also, i was a bit confused by the 2 fields ```IsImplicit``` and ```IsExplicit```. I think they should be opposite of each other but they are implemented separately. Also I could not find a WalkRecord which fills the IsExplicit field.

### CODE REVIEW

**Comment by ddobrev** [comment]
@genuinelucifer I am discussing about IsExplicit/IsImplicit with @tritao . In the mean time, please convert your tabs to 4 spaces and add () to D->isImplicit in Parser.cpp. The latter is not a field but a function, I am pretty sure your code does not compile at the moment. You can also start with the test. We have a special test project for checking the obtained AST, CppSharp.Generator.Tests. You can see how it works in the existing tests.

File: src/Generator.Tests/AST/TestAST.cs [hunk 1]
```diff
@@ -281,5 +281,12 @@ public void TestMacroLineNumber()
         {
             Assert.AreEqual(98, AstContext.FindClass("HasAmbiguousFunctions").First().Specifiers.Last().LineNumberStart);
         }
+
+        [Test]
+        public void TestImplicitDeclaration()
+        {
+            Assert.AreEqual(false, AstContext.FindFunction("testExplicitFunc").Single().IsImplicit);
+            Assert.AreEqual(true, AstContext.FindFunction("testImplicitFunc").Single().IsImplicit);
```
Comment (genuinelucifer) [inline src/Generator.Tests/AST/TestAST.cs]: This test is failing. Am I missing something about Implicit and Explicit declarations?

File: src/CppParser/AST.h [hunk 2]
```diff
@@ -642,9 +643,8 @@ class CS_API Method : public Function
     bool IsVirtual;
     bool IsStatic;
     bool IsConst;
-    bool IsImplicit;
-    bool IsExplicit;
     bool IsOverride;
+    bool IsExplicit;
```
Comment (tritao) [inline src/CppParser/AST.h]: These changes can be reverted.

File: src/Generator.Tests/AST/TestAST.cs [hunk 3]
```diff
@@ -281,5 +281,12 @@ public void TestMacroLineNumber()
         {
             Assert.AreEqual(98, AstContext.FindClass("HasAmbiguousFunctions").First().Specifiers.Last().LineNumberStart);
         }
+
+        [Test]
+        public void TestImplicitDeclaration()
+        {
+            Assert.AreEqual(false, AstContext.FindFunction("testExplicitFunc").Single().IsImplicit);
```
Comment (ddobrev) [inline src/Generator.Tests/AST/TestAST.cs]: Please just ignore IsExplicit. It works correctly at the moment and does not need any related changes.

File: tests/Native/AST.h [hunk 4]
```diff
@@ -106,3 +106,13 @@ class Atomics
 # endif
 #endif
 };
+
```
Comment (ddobrev) [inline tests/Native/AST.h]: I should explain what implicit declarations are. These are declarations which are not written at all but are instead generated by the compiler. The most common examples are the empty constructor, copy constructor, assigment operator and destructor in C++ classes. So in order to test implicity, just select any class without any of these members explicitly defined in code, and assert that IsImplicit is true. For example, let's say you have this class:

class ImplicitCtor
{
};

Then you can do Assert.IsTrue(classes.Find("ImplicitCtor").Constructors.First(c => c.Parameter.Count ==0).IsImplicit);

Comment (genuinelucifer) [inline tests/Native/AST.h]: Thanks. I don't know why but I remembered from somewhere that functions have implicit and explicit declarations (not necessarily in class).
Well, I have fixed it now as you told. Please review.

File: src/Generator.Tests/AST/TestAST.cs [hunk 5]
```diff
@@ -281,5 +281,11 @@ public void TestMacroLineNumber()
         {
             Assert.AreEqual(98, AstContext.FindClass("HasAmbiguousFunctions").First().Specifiers.Last().LineNumberStart);
         }
+
+        [Test]
+        public void TestImplicitDeclaration()
+        {
+            Assert.IsTrue(AstContext.FindClass("ImplicitCtor").First().Constructors.First(c => c.Parameters.Count == 0).IsImplicit);
```
Comment (ddobrev) [inline src/Generator.Tests/AST/TestAST.cs]: Please break this line - after "First(" is a good location - and I'll merge.

**Comment by ddobrev** [comment]
@genuinelucifer thank you for your work. Let me know when you are ready for a new task.

**Comment by genuinelucifer** [comment]
@ddobrev Pls provide me a new task to work on.

**Comment by ddobrev** [comment]
@genuinelucifer since you are more experienced now compared to 3 or 4 months ago, would you liek to try the one for StringBuilder (https://github.com/mono/CppSharp/issues/508) again?

**Comment by genuinelucifer** [comment]
@ddobrev I'll definitely try that one again.  I guess their will be a few
doubts in that issue. I'll try first and then ask you about those.

**Pull request merged**